security-utils/src/main/java/com/yahoo/security/tls/authz/PeerAuthorizer.java

```diff
@@ -0,0 +1,75 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.security.tls.authz;
+
+import com.yahoo.security.SubjectAlternativeName;
+import com.yahoo.security.X509CertificateUtils;
+import com.yahoo.security.tls.policy.AuthorizedPeers;
+import com.yahoo.security.tls.policy.PeerPolicy;
+import com.yahoo.security.tls.policy.RequiredPeerCredential;
+import com.yahoo.security.tls.policy.Role;
+
+import java.security.cert.X509Certificate;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+import static com.yahoo.security.SubjectAlternativeName.Type.DNS_NAME;
+import static com.yahoo.security.SubjectAlternativeName.Type.IP_ADDRESS;
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Uses rules from {@link AuthorizedPeers} to evaluate X509 certificates
+ *
+ * @author bjorncs
+ */
+public class PeerAuthorizer {
+    private final AuthorizedPeers authorizedPeers;
+
+    public PeerAuthorizer(AuthorizedPeers authorizedPeers) {
+        this.authorizedPeers = authorizedPeers;
+    }
+
+    public AuthorizationResult authorizePeer(X509Certificate peerCertificate) {
+        Set<Role> assumedRoles = new HashSet<>();
+        Set<String> matchedPolicies = new HashSet<>();
+        String cn = getCommonName(peerCertificate).orElse(null);
+        List<String> sans = getSubjectAlternativeNames(peerCertificate);
+        for (PeerPolicy peerPolicy : authorizedPeers.peerPolicies()) {
+            if (matchesPolicy(peerPolicy, cn, sans)) {
+                assumedRoles.addAll(peerPolicy.assumedRoles());
+                matchedPolicies.add(peerPolicy.policyName());
+            }
+        }
+        return new AuthorizationResult(assumedRoles, matchedPolicies);
+    }
+
+    private static boolean matchesPolicy(PeerPolicy peerPolicy, String cn, List<String> sans) {
+        return peerPolicy.requiredCredentials().stream()
+                .allMatch(requiredCredential -> matchesRequiredCredentials(requiredCredential, cn, sans));
+    }
+
+    private static boolean matchesRequiredCredentials(RequiredPeerCredential requiredCredential, String cn, List<String> sans) {
+        switch (requiredCredential.field()) {
+            case CN:
+                return cn != null && requiredCredential.pattern().matches(cn);
+            case SAN_DNS:
+                return sans.stream()
+                        .anyMatch(san -> requiredCredential.pattern().matches(san));
+            default:
+                throw new RuntimeException("Unknown field: " + requiredCredential.field());
+        }
+    }
+
+    private static Optional<String> getCommonName(X509Certificate peerCertificate) {
+        return X509CertificateUtils.getSubjectCommonNames(peerCertificate).stream()
+                .findFirst();
```

Comment:
In the face of multiple CNs, RFC 6125 states that the "most specific" CN should be used. Going by e.g. [Boost.Asio's RFC2818 implementation](https://github.com/boostorg/asio/blob/develop/include/boost/asio/ssl/impl/rfc2818_verification.ipp#L104) this looks to be the _last_ entry in the certificate.

... Assuming this method returns CNs in original DER-encoded order. I.e. since Distinguished Name printing _reverses_ the order of the fields in DER form, from that perspective the most specific entry is the _first_ one. This is a bit of a rat's nest of diffuse definitions 😬

Reply:
The current strategy of picking the first should be "correct" given that DN parsing reverses the order of CN fields compared to the DER format.